File: v4/rust/iprange-livedb/src/publication/reservation_verify.rs
//! Exact reservation and prepared-output custody checks.

use std::fs::File;

use crate::{error, file_io};

use super::super::namespace::{regular_identity, Identity, Name, NamespaceError};
use super::super::output::PreparedOutput;
use super::super::reservation::{self, Header, Selected};
use super::{Error, FILE_SIZE};

#[derive(Clone, Copy)]
pub(super) enum ReservationLocation {
    Private,
    Canonical,
}

#[derive(Clone, Copy)]
pub(super) enum OutputLocation {
    Private,
    Main,
}

pub(super) struct Expected<'a> {
    pub(super) identity: Identity,
    pub(super) private_name: &'a Name,
    pub(super) header: Header,
    pub(super) block: usize,
    pub(super) reservation_location: ReservationLocation,
    pub(super) output_location: OutputLocation,
}

pub(super) fn verify(
    file: &File,
    output: &PreparedOutput,
    expected: Expected<'_>,
) -> Result<(), Error> {
    verify_inode(file, output, &expected)?;
    verify_location(output, &expected)?;
    verify_contents(file, expected.header, expected.block)
}

fn verify_inode(
    file: &File,
    output: &PreparedOutput,
    expected: &Expected<'_>,
) -> Result<(), Error> {
    match expected.output_location {
        OutputLocation::Private => output.verify_private(),
        OutputLocation::Main => output.verify_main(),
    }
    .map_err(Error::Output)?;
    let destination = output.attempt.destination();
    let directory = destination.directory();
    directory.verify()?;
    if regular_identity(file, directory.identity())? != expected.identity {
        return Err(NamespaceError::IdentityChanged.into());
    }
    Ok(destination.verify_created(file)?)
}

fn verify_location(output: &PreparedOutput, expected: &Expected<'_>) -> Result<(), Error> {
    let destination = output.attempt.destination();
    let directory = destination.directory();
    let name = match expected.reservation_location {
        ReservationLocation::Private => expected.private_name,
        ReservationLocation::Canonical => destination.coordination(),
    };
    directory.verify_name(name, expected.identity)?;
    if matches!(
        expected.reservation_location,
        ReservationLocation::Canonical
    ) {
        directory.require_absent(expected.private_name)?;
    }
    Ok(())
}

fn verify_contents(file: &File, header: Header, block: usize) -> Result<(), Error> {
    if file.metadata().map_err(error::Error::from)?.len() != FILE_SIZE as u64 {
        return Err(Error::LengthChanged);
    }
    select_exact(file, header, block)
}

pub(super) fn select_exact(file: &File, header: Header, block: usize) -> Result<(), Error> {
    let mut bytes = [0; FILE_SIZE];
    file_io::read_exact_at(file, &mut bytes, 0)?;
    if reservation::select(&bytes).map_err(Error::Codec)? != (Selected { header, block }) {
        return Err(Error::HeaderChanged);
    }
    Ok(())
}

File: v4/go/internal/publication/reservation_verify.go
// Exact reservation and prepared-output custody checks (Rust
// publication/reservation_verify.rs). One verification proves the
// reservation inode against the output evidence, its placement at the
// private or canonical name, and the exact dual-block record; every
// read runs over the mapped reservation view with no intermediate
// copy.

package publication

import (
	"os"

	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/live"
	"github.com/firehol/iprange/v4/go/internal/mapping"
)

// reservationLocation is the namespace position of one reservation
// (Rust ReservationLocation).
type reservationLocation uint8

const (
	reservationLocationPrivate reservationLocation = iota
	reservationLocationCanonical
)

// reservationExpected is the exact custody expectation of one
// reservation verification (Rust Expected).
type reservationExpected struct {
	identity            live.FileIdentity
	privateName         string
	header              reservationHeader
	block               int
	reservationLocation reservationLocation
	outputLocation      outputLocation
}

// verifyReservation runs the three-part custody proof (Rust verify):
// inode evidence, namespace placement, and record contents.
func verifyReservation(file *os.File, m *mapping.Mapping, output *preparedOutput, expected reservationExpected) error {
	if err := verifyReservationInode(file, output, expected); err != nil {
		return err
	}
	if err := verifyReservationLocation(output, expected); err != nil {
		return err
	}
	return verifyReservationContents(file, m, expected.header, expected.block)
}

// verifyReservationInode proves the output is still at its exact
// position, the retained directory is stable, and the reservation
// file is the expected single-link inode with the creator commitment
// (Rust verify_inode).
func verifyReservationInode(file *os.File, output *preparedOutput, expected reservationExpected) error {
	switch expected.outputLocation {
	case outputLocationPrivate:
		if err := output.verifyPrivate(); err != nil {
			return err
		}
	case outputLocationMain:
		if err := output.verifyMain(); err != nil {
			return err
		}
	}
	destination := output.attempt.destinationOf()
	directory := destination.directory()
	if err := directory.Verify(); err != nil {
		return err
	}
	identity, err := live.RegularIdentity(file, directory.Identity())
	if err != nil {
		return err
	}
	if identity != expected.identity {
		return &live.NamespaceError{Kind: live.NamespaceIdentityChanged}
	}
	return destination.verifyCreated(file)
}

// verifyReservationLocation proves the reservation name placement and
// the private-name absence at the canonical position (Rust
// verify_location, including the gc-barrier availability call with
// the reservation artifact kind).
func verifyReservationLocation(output *preparedOutput, expected reservationExpected) error {
	destination := output.attempt.destinationOf()
	directory := destination.directory()
	name := expected.privateName
	kind := ArtifactPrivateReservation
	if expected.reservationLocation == reservationLocationCanonical {
		name = destination.coordinationName()
		kind = ArtifactOwnedCoordination
	}
	if err := requireSourceAvailable(directory, expected.header.attemptID, 1, kind, DirectoryRoleDestination, name, expected.identity); err != nil {
		return err
	}
	if err := directory.VerifyName(name, expected.identity); err != nil {
		return err
	}
	if expected.reservationLocation == reservationLocationCanonical {
		return directory.RequireAbsent(expected.privateName)
	}
	return nil
}

// verifyReservationContents proves the reservation file keeps its
// exact size and its selected record (Rust verify_contents).
func verifyReservationContents(file *os.File, m *mapping.Mapping, header reservationHeader, block int) error {
	size, err := fstatSize(file)
	if err != nil {
		return &format.Error{Code: format.CodeIO, Detail: "stat: " + err.Error()}
	}
	if size != reservationFileSize {
		return reservationLengthChangedError()
	}
	return selectExact(m, header, block)
}

// selectExact proves the mapped reservation still selects the exact
// expected record at its block (Rust select_exact: any select refusal
// is the codec class, any selection mismatch is the header-changed
// class).
func selectExact(m *mapping.Mapping, header reservationHeader, block int) error {
	bytes, err := m.View(0, reservationFileSize)
	if err != nil {
		return err
	}
	selected, err := selectReservation(bytes)
	if err != nil {
		return reservationCodecError()
	}
	if selected.header != header || selected.block != block {
		return reservationHeaderChangedError()
	}
	return nil
}
